File: android/app/src/main/java/com/laneshadow/ui/components/molecules/DownloadErrorSheet.kt
package com.laneshadow.ui.components.molecules

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.laneshadow.ui.components.atoms.Button
import com.laneshadow.ui.components.atoms.ButtonVariant
import com.laneshadow.theme.LocalLaneShadowTheme

/**
 * DownloadErrorSheet molecule component
 *
 * Bottom sheet dialog for download error states with retry and support options.
 * Following React Native wrapper patterns from react-native/components/offline/download-error-sheet.tsx
 *
 * @param isVisible Whether to show the error sheet
 * @param onRetry Callback when user taps retry button
 * @param onClose Callback when user cancels or dismisses the sheet
 * @param error Optional error message to display (defaults to generic message)
 * @param retryCount Number of retry attempts (defaults to 0, shows "Contact Support" at 3+)
 * @param testID Optional test ID for UI testing
 */
@Composable
fun DownloadErrorSheet(
    isVisible: Boolean,
    onRetry: () -> Unit,
    onClose: () -> Unit,
    error: String? = null,
    retryCount: Int = 0,
    testID: String? = null,
) {
    if (!isVisible) {
        return
    }

    val theme = LocalLaneShadowTheme.current

    // Default error message if none provided
    val errorMessage = error ?: "There was a problem downloading this map. Please check your connection and try again."

    // Show contact support option after 3 retries
    val showContactSupport = retryCount >= 3

    // Build accessibility description
    val accessibilityDescription = "Download failed. $errorMessage"

    AlertDialog(
        onDismissRequest = onClose,
        modifier = Modifier
            .testTag(testID ?: "download-error-sheet")
            .semantics {
                contentDescription = accessibilityDescription
            },
        title = null,
        text = {
            Column(
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(horizontal = 24.dp),
                horizontalAlignment = Alignment.CenterHorizontally,
            ) {
                // Error icon (64x64 danger circle with "!" text)
                ErrorIcon(
                    modifier = Modifier
                        .size(64.dp)
                        .testTag(testID?.let { "$it-icon" } ?: "download-error-icon")
                )

                Spacer(modifier = Modifier.height(theme.space.lg))

                // Title: "Download Failed" (title.lg = 18sp, SemiBold)
                Text(
                    text = "Download Failed",
                    color = theme.colors.onSurface.default,
                    style = androidx.compose.ui.text.TextStyle(
                        fontSize = 18.sp,
                        fontWeight = FontWeight.SemiBold,
                    ),
                    textAlign = TextAlign.Center,
                    modifier = Modifier
                        .fillMaxWidth()
                        .testTag(testID?.let { "$it-title" } ?: "download-error-title")
                )

                Spacer(modifier = Modifier.height(theme.space.md))

                // Error message (body.md = 14sp, muted color)
                Text(
                    text = errorMessage,
                    color = theme.colors.onSurface.default.copy(alpha = 0.6f),
                    style = androidx.compose.ui.text.TextStyle(
                        fontSize = 14.sp,
                        fontWeight = FontWeight.Normal,
                    ),
                    textAlign = TextAlign.Center,
                    modifier = Modifier
                        .fillMaxWidth()
                        .testTag(testID?.let { "$it-message" } ?: "download-error-message")
                )

                Spacer(modifier = Modifier.height(theme.space.lg))

                // Retry Download button (default variant, lg size = 44dp, full width)
                Button(
                    variant = ButtonVariant.Default,
                    size = com.laneshadow.ui.components.atoms.ButtonSize.Large,
                    text = "Retry Download",
                    onPress = onRetry,
                    modifier = Modifier
                        .fillMaxWidth()
                        .testTag(testID?.let { "$it-retry" } ?: "download-error-retry"),
                )

                Spacer(modifier = Modifier.height(theme.space.md))

                // Contact Support button (ghost, shown when retryCount >= 3)
                if (showContactSupport) {
                    Button(
                        variant = ButtonVariant.Ghost,
                        size = com.laneshadow.ui.components.atoms.ButtonSize.Default,
                        text = "Contact Support",
                        onPress = {
                            // TODO: Navigate to support flow
                            onClose()
                        },
                        modifier = Modifier
                            .fillMaxWidth()
                            .testTag(testID?.let { "$it-support" } ?: "download-error-support"),
                    )

                    Spacer(modifier = Modifier.height(theme.space.md))
                }

                // Cancel button (ghost, full width)
                Button(
                    variant = ButtonVariant.Ghost,
                    size = com.laneshadow.ui.components.atoms.ButtonSize.Default,
                    text = "Cancel",
                    onPress = onClose,
                    modifier = Modifier
                        .fillMaxWidth()
                        .testTag(testID?.let { "$it-cancel" } ?: "download-error-cancel"),
                )
            }
        },
        confirmButton = {},
        containerColor = theme.colors.surface.default,
        shape = androidx.compose.foundation.shape.RoundedCornerShape(theme.radius.xl),
    )
}

/**
 * Error icon component
 *
 * 64x64 danger circle background with "!" text in center.
 * Following RN wrapper design tokens.
 *
 * @param modifier Modifier for the icon
 */
@Composable
private fun ErrorIcon(
    modifier: Modifier = Modifier,
) {
    val theme = LocalLaneShadowTheme.current

    // Circular background with danger color
    Box(
        modifier = modifier
            .background(
                color = theme.colors.danger.default,
                shape = CircleShape,
            ),
        contentAlignment = Alignment.Center,
    ) {
        // Exclamation mark text
        Text(
            text = "!",
            color = theme.colors.onPrimary.default,
            style = androidx.compose.ui.text.TextStyle(
                fontSize = 32.sp,
                fontWeight = FontWeight.Bold,
            ),
        )
    }
}

File: ios/LaneShadow/Views/Molecules/DownloadErrorSheet.swift
import LaneShadowTheme
import SwiftUI

// MARK: - DownloadErrorSheet Component

/**
 * DownloadErrorSheet molecule component
 *
 * Bottom sheet dialog for download error states with retry and support options.
 * Following React Native component from react-native/components/offline/download-error-sheet.tsx
 *
 * ## Design Tokens Used
 * - Colors:
 *   - Background: `theme.colors.surface.default`
 *   - Error icon background: `theme.colors.danger.default`
 *   - Error icon text: `theme.colors.onPrimary.default`
 *   - Title: `theme.colors.onSurface.default`
 *   - Message: `theme.colors.onSurface.default` with 0.6 opacity
 * - Layout:
 *   - Content padding: 24pt
 *   - Gap: `theme.space.lg` (16pt)
 *   - Error icon size: 64pt
 * - Typography:
 *   - Title: 18pt semibold (title.lg)
 *   - Message: 14pt regular (body.md)
 *   - Error icon: 32pt bold
 *
 * ## Parameters
 * - isVisible: Whether to show the error sheet
 * - onRetry: Callback when user taps retry button
 * - onClose: Callback when user cancels or dismisses the sheet
 * - error: Optional error message to display (defaults to generic message)
 * - retryCount: Number of retry attempts (defaults to 0, shows "Contact Support" at 3+)
 * - testID: Optional test ID for UI testing
 */
public struct LSDownloadErrorSheet: View {
    @Environment(\.theme) private var theme

    private let isVisible: Bool
    private let onRetry: () -> Void
    private let onClose: () -> Void
    private let error: String?
    private let retryCount: Int
    private let testID: String?

    private let defaultErrorMessage = "There was a problem downloading this map. Please check your connection and try again."

    public init(
        isVisible: Bool,
        onRetry: @escaping () -> Void,
        onClose: @escaping () -> Void,
        error: String? = nil,
        retryCount: Int = 0,
        testID: String? = nil
    ) {
        self.isVisible = isVisible
        self.onRetry = onRetry
        self.onClose = onClose
        self.error = error
        self.retryCount = retryCount
        self.testID = testID
    }

    // MARK: - Body

    public var body: some View {
        EmptyView()
            .sheet(isPresented: .constant(isVisible)) {
                sheetContent
            }
            .accessibilityIdentifier(testID ?? "download-error-sheet")
    }

    // MARK: - Sheet Content

    private var sheetContent: some View {
        VStack(spacing: 0) {
            // Drag indicator for bottom sheet
            RoundedRectangle(cornerRadius: 2)
                .fill(theme.colors.onSurface.default.opacity(0.3))
                .frame(width: 36, height: 4)
                .padding(.top, theme.space.md)
                .accessibilityHidden(true)

            VStack(spacing: theme.space.lg) {
                // Error icon (64x64 danger circle with "!" text)
                errorIcon
                    .accessibilityIdentifier(testID?.let { "\($0)-icon" } ?? "download-error-icon")

                // Title: "Download Failed"
                Text("Download Failed")
                    .font(.system(size: 18, weight: .semibold))
                    .foregroundStyle(theme.colors.onSurface.default)
                    .multilineTextAlignment(.center)
                    .accessibilityIdentifier(testID?.let { "\($0)-title" } ?? "download-error-title")

                // Error message
                Text(errorMessage)
                    .font(.system(size: 14, weight: .regular))
                    .foregroundStyle(theme.colors.onSurface.default.opacity(0.6))
                    .multilineTextAlignment(.center)
                    .accessibilityIdentifier(testID?.let { "\($0)-message" } ?? "download-error-message")

                // Retry Download button (default variant, lg size, full width)
                LSButton(
                    "Retry Download",
                    variant: .default,
                    size: .lg,
                    onPress: onRetry
                )
                .accessibilityIdentifier(testID?.let { "\($0)-retry" } ?? "download-error-retry")

                VStack(spacing: theme.space.md) {
                    // Contact Support button (ghost, shown when retryCount >= 3)
                    if showContactSupport {
                        LSButton(
                            "Contact Support",
                            variant: .ghost,
                            size: .default,
                            onPress: {
                                // TODO: Navigate to support flow
                                onClose()
                            }
                        )
                        .accessibilityIdentifier(testID?.let { "\($0)-support" } ?? "download-error-support")
                    }

                    // Cancel button (ghost, full width)
                    LSButton(
                        "Cancel",
                        variant: .ghost,
                        size: .default,
                        onPress: onClose
                    )
                    .accessibilityIdentifier(testID?.let { "\($0)-cancel" } ?? "download-error-cancel")
                }
            }
            .padding(.horizontal, 24)
            .padding(.vertical, theme.space.lg)
        }
        .background(theme.colors.surface.default)
        .clipShape(RoundedRectangle(cornerRadius: theme.radius.xl))
        .presentationDetents([.large])
        .presentationDragIndicator(.hidden)
        .accessibilityElement(children: .contain)
        .accessibilityLabel("Download failed. \(errorMessage)")
    }

    // MARK: - Error Icon

    private var errorIcon: some View {
        ZStack {
            Circle()
                .fill(theme.colors.danger.default)
                .frame(width: 64, height: 64)

            Text("!")
                .font(.system(size: 32, weight: .bold))
                .foregroundStyle(theme.colors.onPrimary.default)
        }
    }

    // MARK: - Computed Properties

    private var errorMessage: String {
        error ?? defaultErrorMessage
    }

    private var showContactSupport: Bool {
        retryCount >= 3
    }
}

// MARK: - Preview

#Preview("DownloadErrorSheet - Default") {
    LSDownloadErrorSheet(
        isVisible: true,
        onRetry: {
            print("Retry tapped")
        },
        onClose: {
            print("Close tapped")
        },
        testID: "preview-download-error"
    )
    .laneShadowTheme()
}

#Preview("DownloadErrorSheet - Custom Error") {
    LSDownloadErrorSheet(
        isVisible: true,
        error: "Network connection failed. Please check your internet connection.",
        retryCount: 1,
        onRetry: {
            print("Retry tapped")
        },
        onClose: {
            print("Close tapped")
        },
        testID: "preview-custom-error"
    )
    .laneShadowTheme()
}

#Preview("DownloadErrorSheet - With Contact Support") {
    LSDownloadErrorSheet(
        isVisible: true,
        error: "Download failed after multiple attempts",
        retryCount: 4,
        onRetry: {
            print("Retry tapped")
        },
        onClose: {
            print("Close tapped")
        },
        testID: "preview-with-support"
    )
    .laneShadowTheme()
}

#Preview("DownloadErrorSheet - Hidden") {
    VStack {
        Text("Parent View")
            .font(.headline)
            .padding()

        LSDownloadErrorSheet(
            isVisible: false,
            onRetry: {
                print("Retry tapped")
            },
            onClose: {
                print("Close tapped")
            }
        )
    }
    .laneShadowTheme()
}
